File: Assets/InsightARWorld/InsightARExporter/Utility/BuiltinLuaScripts/Landmark/LandmarkTrackState.lua

local LandmarkTrackState = {}


function LandmarkTrackState:New(game_object)
    if self ~= LandmarkTrackState then
        return nil, "LandmarkTrackState:New is nil"
    end
    local newinstance = setmetatable({}, {__metatable = {}, __index = LandmarkTrackState})
    newinstance.transform = game_object.transform
    transform = game_object.transform
    newinstance.gameObject = game_object
    gameObject = game_object
    return newinstance
end

function LandmarkTrackState:Awake()
	
	
end

function LandmarkTrackState:Start()
	
end

function LandmarkTrackState:Enter()
	Insight.Debug.Log("land mark track state");
	self:Init();
end

function LandmarkTrackState:Execute()
	if(not self.isEnabled) then
		return;
	end
	
	local state = self:CheckTrackState();
	self:UpdateView(state);
end

function LandmarkTrackState:Exit()
	self:Close();
end

function LandmarkTrackState:Init()
	self.isEnabled = true;
	g_ResManager.trackPanelTrans.gameObject:SetActive(true);
	
	-- 开启模型显示
	g_ResManager.modelRootTrans.gameObject:SetActive(true);
	-- g_ResManager.jinyujiTrans.gameObject:SetActive(true);
	-- g_ResManager.datiangouTrans.gameObject:SetActive(true);
	-- g_ResManager.yijinzhentianTrans.gameObject:SetActive(true);
	-- g_ResManager.jinlongTrans.gameObject:SetActive(true);
	-- g_ResManager.balloonTrans.gameObject:SetActive(true);
	-- g_ResManager.landmarkTrans.gameObject:SetActive(true);
	-- g_ResManager.axisTrans.gameObject:SetActive(true);
	-- g_ResManager.maskTrans.gameObject:SetActive(true);
end


function LandmarkTrackState:Close()
	self.isEnabled = false;
	g_ResManager.trackPanelTrans.gameObject:SetActive(false);
	--关闭模型显示
	g_ResManager.modelRootTrans.gameObject:SetActive(false);
	-- g_ResManager.jinyujiTrans.gameObject:SetActive(false);
	-- g_ResManager.datiangouTrans.gameObject:SetActive(false);
	-- g_ResManager.yijinzhentianTrans.gameObject:SetActive(false);
	-- g_ResManager.jinlongTrans.gameObject:SetActive(false);
	-- g_ResManager.balloonTrans.gameObject:SetActive(false);
	-- g_ResManager.landmarkTrans.gameObject:SetActive(false);
	-- g_ResManager.axisTrans.gameObject:SetActive(false);
	-- g_ResManager.maskTrans.gameObject:SetActive(false);
end

-- 更新视图
function LandmarkTrackState:UpdateView(state)
	if(state == 0) then
		if(not self.mapEnabled) then
			self.mapEnabled = true;
			
			g_ResManager.downIconTrans.gameObject:SetActive(false);
			g_ResManager.upIconTrans.gameObject:SetActive(false);
			g_ResManager.trackLimitedTrans.gameObject:SetActive(false);
			
		end
	else
		if(self.mapEnabled) then
			self.mapEnabled = false;

			if(state == 1) then
				g_ResManager.downIconTrans.gameObject:SetActive(true);
			elseif(state == 2) then
				g_ResManager.upIconTrans.gameObject:SetActive(true);
			elseif(state == 3) then
				g_ResManager.trackLimitedTrans.gameObject:SetActive(true);
			end
		end
	end
end

--检查当前是否处于用户操作异常状态
function LandmarkTrackState:CheckTrackState()
	if(Insight.Tracking.status == 7) then
		local upVect = Insight.Vector3.New(0,1,0);
		local downVect = Insight.Vector3.New(0,-1,0);
		local upDot = Insight.Vector3.Dot(g_ResManager.mainCameraTrans.forward,upVect);
		local downDot = Insight.Vector3.Dot(g_ResManager.mainCameraTrans.forward,downVect);
			--阈值可以调节
		if(upDot >0.9 and downDot <0)then
			-- 手机朝着天空需要下移
			return 1;
		end
			
		if(downDot >0.9 and upDot <0 ) then
			-- 手机朝着地面需要上抬
			return 2;
		end
		--正常
		return 0;
	else
		--算法状态异常
		return 3;
	end
end

return LandmarkTrackState

File: U3DExporter/2022_10_29_10_04_55_Android/Assets/InsightARWorld/InsightARExporter/Utility/BuiltinJSScripts/Landmark/LandmarkTrackState.js
var g_lastRecognitionCount = 0;
var g_lastErrorCode = null;
var g_landmarkTrackStateInstance = null;

function LandmarkTrackState(gameObject)
{
    this.transform = gameObject.transform;
    this.gameObject = gameObject;
}
//Write prototype function here
LandmarkTrackState.prototype = Object.assign(Object.create(Object.prototype), {
    // Start is called before the first frame update
    Start: function(){
        g_landmarkTrackStateInstance = this;
    },
    Awake: function(){

    },
    // Update is called once per frame
    Update: function()
    {
        
    },
    Enter: function()
	{
        Insight.Debug.Log("land mark track state");
        this.Init();
    },
    Execute: function()
    {
        if(!this.isEnabled)
        {
		    return;
        }
	
        var state = this.CheckTrackState();
        this.UpdateView(state);
    },
    Exit: function()
	{
        this.Close();
        g_ResManager.trackPanelTrans.gameObject.setActive(false);
        //关闭模型显示
        g_ResManager.modelRootTrans.gameObject.setActive(false);
    },
    Init: function()
    {
        this.isEnabled = true;
        this.mapEnabled = false;
        this.limitFailCount = 5;
        g_ResManager.trackPanelTrans.gameObject.setActive(true);
        
        // 开启模型显示
        g_ResManager.modelRootTrans.gameObject.setActive(true);
    },
    Close: function()
    {
        this.isEnabled = false;    
    },
    // 更新视图
    UpdateView: function(state)
    {
        if(state == 0)
        {
            if(!this.mapEnabled)
            {
                this.mapEnabled = true;
                
                g_ResManager.downIconTrans.gameObject.setActive(false);
                g_ResManager.upIconTrans.gameObject.setActive(false);
                g_ResManager.trackLimitedTrans.gameObject.setActive(false);   
                g_ResManager.trackFailTrans.gameObject.setActive(false);
                g_ResManager.errorCodeTrans.gameObject.setActive(false);
            }
        }
        else
        {
            if(state==null) return;

            if(this.mapEnabled)
            {
                this.mapEnabled = false;

                if(state == 1)
                {
                    g_ResManager.downIconTrans.gameObject.setActive(true);
                    g_ResManager.upIconTrans.gameObject.setActive(false);
                    g_ResManager.trackLimitedTrans.gameObject.setActive(false);
                    g_ResManager.trackFailTrans.gameObject.setActive(false);
                    g_ResManager.errorCodeTrans.gameObject.setActive(false);
                }
                else if(state == 2)
                {
                    g_ResManager.upIconTrans.gameObject.setActive(true);
                    g_ResManager.downIconTrans.gameObject.setActive(false);
                    g_ResManager.trackLimitedTrans.gameObject.setActive(false);
                    g_ResManager.trackFailTrans.gameObject.setActive(false);
                    g_ResManager.errorCodeTrans.gameObject.setActive(false);
                }
                else
                {

                }
            }
        }
    },
    //检查当前是否处于用户操作异常状态
    CheckTrackState: function()
    {
         if(Insight.Tracking.cloudLocationStatus == 1)
        {
            var upVect = new Insight.Vector3(0, 1, 0);//Insight.Vector3.New(0,1,0);
            var downVect = new Insight.Vector3(0, -1, 0);//Insight.Vector3.New(0,-1,0);
            var upDot = g_ResManager.mainCameraTrans.forward.dot(upVect);//Insight.Vector3.Dot(g_ResManager.mainCameraTrans.forward,upVect);
            var downDot = g_ResManager.mainCameraTrans.forward.dot(downVect);//Insight.Vector3.Dot(g_ResManager.mainCameraTrans.forward,downVect);
            //阈值可以调节
            if(upDot >0.9 && downDot <0)
            {
                // 手机朝着天空需要下移
                return 1;
            }
                
            if(downDot >0.9 && upDot <0 )
            {
                // 手机朝着地面需要上抬
                return 2;
            }
            //正常
            return 0;
        }
    }
});

//Return the script module
LandmarkTrackState